### PR TITLE
Separate commands for toggling status and message buffers

### PR DESCRIPTION
**Pull request by guojing0**
This is to resolve Issue https://github.com/atlas-engineer/nyxt/issues/2009 by introducing `toggle-status-buffer` and `toggle-message-buffer` commands.

### CODE REVIEW

**Comment by aartaka** [comment]
CI breaks because removed `(un)present-current-window` is used elsewhere :)

A question to test the use-case: what if I want to make one command call to disable the status buffer? Even if it's disabled. What should I do then? That's why I suggested adding `enable-p` keyword argument to the `toggle-status-bar` in #2009: one may want to disable things unconditionally of thier current state.

File: source/window.lisp [hunk 1]
```diff
@@ -237,20 +237,19 @@ mapped to query parameters."
...
+(define-command toggle-status-buffer (&optional (window (current-window)))
+  "Toggle the visibility of the status buffer."
+  (if (zerop (ffi-window-get-status-buffer-height window))
+      (ffi-window-set-status-buffer-height window (height (status-buffer window)))
+      (ffi-window-set-status-buffer-height window 0)))
+
+(define-command toggle-message-buffer (&optional (window (current-window)))
+  "Toggle the visibility of the message buffer."
+  (if (zerop (ffi-window-get-message-buffer-height window))
+      (ffi-window-set-message-buffer-height window (height (message-buffer-height window)))
```
Comment (aartaka) [inline source/window.lisp:254]: What does `(height (message-buffer-height window))` returns?

**Comment by aartaka** [comment]
To rephrase my chaotic questions, here's a list of distinct API actions one should have with status bar and message area:
- Enable status bar.
- Disable status bar.
- Toggle status bar.
- Enable message area.
- Disable message area.
- Toggle message area.
- Enable both.
- Disable both.
- Toggle both.

The most ergonomic way that I can think of is three commands with keyword arguments from #2009, but I may be mistaken.

**Comment by guojing0** [comment]
I vaguely remember that we may have a macro that automatically generates `enable-*`, `disable-*`, and `toggle-*` commands?

**Comment by aartaka** [comment]
> I vaguely remember that we may have a macro that automatically generates `enable-*`, `disable-*`, and `toggle-*` commands?

I don't remember of such a macro, but we don't need it, actually: we only need some three commands and several helper functions.

**Comment by aartaka** [comment]
The way I'd do it is close to what I suggested in #2009 :)

**Comment by guojing0** [comment]
> what if I want to make one command call to disable the status buffer? Even if it's disabled already.

Sorry I do not follow. Why would one do that?

**Comment by aartaka** [comment]
> > what if I want to make one command call to disable the status buffer? Even if it's disabled already.
> 
> Sorry I do not follow. Why would one do that?

This might be useful when writing configuration snippets and to replace `(un)present-current-window`.
- Existing use-case: we need to unconditionally hide both aread when fullscreening a video and unconditionally show those when unfullscreening.
- Future use-case: one might want to have a minimum distraction mode where it unconditionally disables everything but the web page. Even if other mode have already disabled e.g. status buffer, it should still keep it disabled, instead of toggling.

> I agree with `enable-p`. To understand the desired behavior of `enable-p`, if it's set to `T`, then that area is hidden?

If set to `T`, then show; if set to `NIL`, then hide; if not provided, toggle. Actually, it might be better to name it `show-p`. What do you think?

**Comment by guojing0** [comment]
Thank you for the explanation, Artyom. It's really useful to understand better.

However, given current implementation, I do not think we need that many APIs. Say, if an user only wants to permanently disable status buffer, then he can just add something like `(toggle-status-buffer :show-p nil)` in his config file.

File: source/window.lisp [hunk 2]
```diff
@@ -239,18 +239,57 @@ mapped to query parameters."
...
+(defun disable-status-buffer (&optional (window (current-window)))
+  (ffi-window-set-status-buffer-height window 0))
+
+(defun enable-message-buffer (&optional (window (current-window)))
+  (ffi-window-set-message-buffer-height window (message-buffer-height window)))
+
+(defun disable-message-buffer (&optional (window (current-window)))
   (ffi-window-set-message-buffer-height window 0))
 
+(define-command toggle-status-buffer (&optional (window (current-window))
```
Comment (aartaka) [inline source/window.lisp:257]: SBCL complains about `&optional` and `&key` arguments being present in the same arglist. Mixing the two confuses both human readers and compilers. 

In this particular case, `&key` would work for both args.

File: source/window.lisp [hunk 3]
```diff
@@ -239,18 +239,57 @@ mapped to query parameters."
...
+                                      &key (show-p nil show-provided-p))
+  "Toggle the visibility of the status buffer by default.
```
Comment (aartaka) [inline source/window.lisp:259]: The "by default" is not necessary here, given that you explain `show-p` later.

File: source/window.lisp [hunk 4]
```diff
@@ -239,18 +239,57 @@ mapped to query parameters."
...
+
+If SHOW-P is provided:
+- If SHOW-P is T, then `status-buffer' is always enabled;
```
Comment (aartaka) [inline source/window.lisp:262]: "If SHOW-P is non-nil", maybe?

Then the second clause becomes "If SHOW-P is nil...".

File: source/window.lisp [hunk 5]
```diff
@@ -239,18 +239,57 @@ mapped to query parameters."
...
+- Otherwise, it is always disabled."
+  (if show-provided-p
```
Comment (aartaka) [inline source/window.lisp:264]: These nested `if` don't look right. How about a `cond`? Something like:
``` lisp
(cond
  ((and show-provided-p show-p)
   (enable-status-buffer window))
  ((and (not show-provided-p)
        (zerop (ffi-window-get-status-buffer-height window)))
   (enable-status-buffer window))
  (t (disable-status-buffer window)))
```

File: source/window.lisp [hunk 6]
```diff
@@ -239,18 +239,57 @@ mapped to query parameters."
...
+      (if show-p
+          (enable-status-buffer window)
+          (disable-status-buffer window))
+      (if (zerop (ffi-window-get-status-buffer-height window))
+          (enable-status-buffer window)
+          (disable-status-buffer window))))
+
+(define-command toggle-message-buffer (&optional (window (current-window))
+                                       &key (show-p nil show-provided-p))
+  "Toggle the visibility of the message buffer by default.
```
Comment (aartaka) [inline source/window.lisp:274]: Same as above.

**Review by aartaka — CHANGES_REQUESTED** [review]
Looks good and comprehensive enough! There are small things that I'd like to see resolved before we merge, though :)

**Comment by aartaka** [comment]
> However, given current implementation, I do not think we need that many APIs, unless I misunderstood your intention: Say, if an user only wants to permanently disable status buffer, then he can just add something like `(toggle-status-buffer :show-p nil)` in his config file.

That's exactly what I meant -- adding command behavior altering keyword arguments qualifies as API expansion :)

**Comment by guojing0** [comment]
Many thanks. In this case, do you think `(un)present-current-window` can be removed? Or, you have some suggestions for a better name, as someone whose second language is English, when I first saw this it's quite confusing.

File: source/window.lisp (hunk 4 above)
Comment (guojing0) [reply source/window.lisp:262]: From the language perspective, "non-nil" is more correct, but would it be more clear if we "only allow" `T` and `NIL`?

**Comment by aartaka** [comment]
> Many thanks. In this case, do you think `(un)present-current-window` can be removed? Or, you have some suggestions for a better name, as someone whose second language is English, when I first saw this it's quite confusing.

I was confused too :D 

Yes, I believe it can be removed. Just use the matching `toggle-message-buffer` and `toggle-status-buffer` instead in `"fullscreen"` and `"unfullscreen"` signals in `renderer/gtk.lisp`

**Pull request closed without merging**